ray-curator/ray_curator/stages/text/embedders/base.py

```diff
@@ -0,0 +1,142 @@
+        """Load the model for inference."""
+        self.model = AutoModel.from_pretrained(self.model_identifier, local_files_only=True)
+        self.model.eval()
+        self.model.to("cuda")
+
+    def process_model_output(
+        self, outputs: torch.Tensor, model_input_batch: dict[str, torch.Tensor] | None = None
+    ) -> torch.Tensor:
+        """Process model outputs to create embeddings."""
+        if self.pooling == "mean_pooling":
+            return self._mean_pooling(outputs, model_input_batch[ATTENTION_MASK_COLUMN])
+        else:
+            return self._get_last_token(outputs, model_input_batch[ATTENTION_MASK_COLUMN])
+
+    def collect_outputs(self, processed_outputs: list[torch.Tensor]) -> cp.ndarray:
+        """Collect embeddings into a cupy array."""
+        # TODO : benchmarking this and maybe stay in cpu land
+        cupy_array_embeddings = [cp.asarray(emb_chunk) for emb_chunk in processed_outputs]
+        return cp.concatenate(cupy_array_embeddings, axis=0)
+
+    def create_output_dataframe(self, df_cpu: pd.DataFrame, collected_output: cp.ndarray) -> pd.DataFrame:
+        """Create output dataframe with embeddings."""
+        # TODO: Consider if it even makes sense to goto cudf or just concat in numpy
+        df_gpu = cudf.DataFrame(index=df_cpu.index)
+        df_gpu[self.embedding_field] = create_list_series_from_1d_or_2d_ar(collected_output, index=df_gpu.index)
+        # Add embedding_field back to cpu dataframe
+        df_cpu[self.embedding_field] = df_gpu.to_pandas()[self.embedding_field]
+        del df_gpu
+        return df_cpu
+
+    def _mean_pooling(self, model_output: torch.Tensor, attention_mask: torch.Tensor) -> torch.Tensor:
+        token_embeddings = model_output[0]
+        # Mask out irrelevant tokens directly without expanding the mask
+        masked_embeddings = token_embeddings.masked_fill(attention_mask.unsqueeze(-1) == 0, 0.0)
+        sum_embeddings = masked_embeddings.sum(dim=1)
+        sum_mask = attention_mask.sum(dim=1, keepdim=True).clamp(min=1e-9)
+        return F.normalize(sum_embeddings / sum_mask, dim=1)
+
+    def _get_last_token(self, model_output: torch.Tensor, attention_mask: torch.Tensor) -> torch.Tensor:
+        token_embeddings = model_output[0]
+        # Get indices of last non-padded tokens for each sequence in batch
+        last_token_indices = attention_mask.sum(dim=1) - 1  # -1 for 0-based indexing
+        last_token_indices = last_token_indices.to(torch.long)  # Ensure indices are of type long
+        batch_size = attention_mask.size(0)
+        batch_indices = torch.arange(batch_size, device=attention_mask.device)
+        # Get embeddings of last non-padded tokens
+        last_token_embeddings = token_embeddings[batch_indices, last_token_indices]
+        return F.normalize(last_token_embeddings, dim=1)
+
+
+@dataclass(kw_only=True)
+class DistributedEmbeddingModelStage(CompositeStage[DocumentBatch, DocumentBatch]):
```

Comment:
Nit: could we keep calling this `EmbeddingCreator`? Or `EmbeddingCreatorStage` if you prefer.